### PR TITLE
Make handle numbering random

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

**Comment by tiziano88** [comment]
I'm not convinced about introducing this nondeterminism (yet).

I suppose having random handles (with sufficient entropy, i.e. as long as they are unguessable) would allow using them as capabilities without having the TCB track their explicit transfer between nodes. 

Are you thinking of removing the separate handle space when sending data over channels, and just rely on handles being embedded as data in the messages? That seems risky to me, and would prevent us from doing other optimisations (e.g. garbage collection, or fine-grain capability tracking).

If instead you intend to still have the TCB keep track of handles that are usable by each node, then randomising them does not really add anything. Obviously in this case the numbers would need to be local to each node (like file descriptors), and be remapped / reassigned on the other side of the channel.

**Comment by daviddrysdale** [comment]
There's definitely no intent to shift from a per-Node numbering space to a global numbering space;  shifting from TCB-mediated capabilities to (effectively) bearer capabilities would be a bad idea.

There's a couple of reasons for randomizing the values within the per-Node numbering space (although neither are particularly compelling, so this is just a draft PR).

The first is to make sure that Nodes don't start making assumptions about the ordering of handle values, which would make it harder to make changes to the runtime in future.

The other motivation arose from a conversation with @benlaurie; we were thinking about a Node that combined code of different levels of provenance.  For example, it might include a well-known & verified deserialization library that converts bytes-and-handles back to Rust types, together with a less-verified chunk of code that uses those types.  For the latter, the combination of knowing that it a) respects the Rust type system (no `unsafe` code) and b) cannot guess the underlying channel handles (this PR) would allow confidence that it can't do some categories of dubious things, even though the code is less-verified.

**Comment by tiziano88** [comment]
Thanks for the details @daviddrysdale . So IIUC this is a "defense in depth" approach by which handle numbers *are* indeed bearer capabilities *within the same node*, which offers some advantages when mixing code with varying levels of trustworthiness. IMO though this is something that should be solved at the source language level (e.g. Rust), since there are other mechanisms (e.g. type systems) by which this confinement of untrusted code can be achieved. From the point of view of the Oak Runtime, the code in a node is all equally trusted, so personally I think this change would make things more confusing, but I could be convinced if you believe it's a good idea.

**Comment by daviddrysdale** [comment]
Leaving aside the second reason for this change, I still think the first reason makes it worthwhile:  Nodes should absolutely not rely on the particular values of handles, and randomizing them ensures that they don't (c.f how Golang explicitly randomizes the order of `map` iteration).

File: sdk/rust/oak_tests/src/lib.rs [hunk 1]
```diff
@@ -413,16 +410,27 @@ impl OakRuntime {
...
     fn new() -> OakNode {
         OakNode {
-            next_handle: 1 as oak::Handle,
             contents_name: "<default>".to_string(),
             halves: HashMap::new(),
             ports: HashMap::new(),
             thread_handle: None,
         }
     }
+    fn next_handle(&self) -> oak::Handle {
```
Comment (tiziano88) [inline sdk/rust/oak_tests/src/lib.rs]: This does not seem thread safe (assuming we wanted this to at some point be promoted to be an "actual" runtime): two concurrent calls may (unlikely) result in the same handle being returned. We should probably add a mutex / RWlock on the handles table?

Comment (daviddrysdale) [reply sdk/rust/oak_tests/src/lib.rs]: I don't think there can be concurrent calls to this &ndash; each `OakNode` object is owned by an `OakRuntime` object, and access to that is already under a `RWLock`.

File: examples/abitest/module/rust/src/lib.rs [hunk 2]
```diff
@@ -236,7 +237,8 @@ impl FrontendNode {
 
     fn test_channel_create(&self) -> std::io::Result<()> {
         let mut handles = Vec::<(oak::WriteHandle, oak::ReadHandle)>::new();
-        for _i in 0..500 {
+        const CHANNEL_COUNT: i32 = 50;
+        for _i in 0..CHANNEL_COUNT {
```
Comment (tiziano88) [inline examples/abitest/module/rust/src/lib.rs]:
```suggestion
        for _ in 0..CHANNEL_COUNT {
```

File: examples/abitest/module/rust/src/lib.rs [hunk 3]
```diff
@@ -236,7 +237,8 @@ impl FrontendNode {
 
     fn test_channel_create(&self) -> std::io::Result<()> {
         let mut handles = Vec::<(oak::WriteHandle, oak::ReadHandle)>::new();
-        for _i in 0..500 {
+        const CHANNEL_COUNT: i32 = 50;
```
Comment (tiziano88) [inline examples/abitest/module/rust/src/lib.rs]:
```suggestion
        const CHANNEL_COUNT: usize = 50;
```

Comment (daviddrysdale) [reply examples/abitest/module/rust/src/lib.rs]: Done.

File: examples/abitest/module/rust/src/lib.rs (hunk 2 above)
Comment (daviddrysdale) [reply examples/abitest/module/rust/src/lib.rs]: Done (and elsewhere)

**Pull request merged**